util/src/RawMemoryPool.cpp

```diff
@@ -0,0 +1,142 @@
+//
+// This product may include a number of subcomponents with separate copyright notices and license terms. Your use of
+// these subcomponents is subject to the terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#include "RawMemoryPool.hpp"
+
+#include <thread>
+#include "assertUtils.hpp"
+#include "kvstream.h"
+
+namespace concordUtil {
+
+using namespace std;
+using namespace std::chrono;
+
+RawMemoryPool::RawMemoryPool(uint32_t chunkSize, Timers& timers) : chunkSize_(chunkSize), metrics_(timers) {}
+
+RawMemoryPool::~RawMemoryPool() {
+  stopWorking_ = true;
+  unique_lock<mutex> lock(waitForAvailChunkLock_);
+  this_thread::sleep_for(std::chrono::milliseconds(WAIT_TIMEOUT_MILLI * 2));
+  char* chunk = nullptr;
+  while (pool_->pop(chunk)) delete[] chunk;
+}
+
+void RawMemoryPool::allocatePool(int32_t initialChunksNum, int32_t maxChunksNum) {
+  if (initialChunksNum > maxChunksNum || initialChunksNum < 0 || maxChunksNum < 0) {
+    stringstream err;
+    err << "Wrong parameters specified: initialChunksNum=" << initialChunksNum << ", maxChunksNum=" << maxChunksNum;
+    LOG_ERROR(logger(), KVLOG(err.str()));
+    throw std::invalid_argument(__PRETTY_FUNCTION__ + err.str());
+  }
+  initialChunksNum_ = initialChunksNum;
+  maxChunksNum_ = maxChunksNum;
+  pool_ = make_shared<boost::lockfree::queue<char*, boost::lockfree::fixed_sized<true>>>(maxChunksNum_);
+  for (int32_t i = 0; i < initialChunksNum_; ++i) {
+    pool_->push(new char[chunkSize_]);
+    metrics_.increaseAvailableChunksNum();
+    metrics_.increaseAllocatedChunksNum();
+  }
+  numOfAvailableChunks_ = initialChunksNum;
+  numOfAllocatedChunks_ = initialChunksNum;
+  LOG_INFO(logger(),
+           "Memory pool allocated" << KVLOG(chunkSize_,
+                                            initialChunksNum_,
+                                            maxChunksNum_,
+                                            numOfAllocatedChunks_,
+                                            numOfAvailableChunks_,
+                                            maxAvailableChunksPercentage_));
+}
+
+void RawMemoryPool::increaseNumOfAllocatedChunks() {
+  numOfAllocatedChunks_++;
+  metrics_.increaseAllocatedChunksNum();
+}
+
+void RawMemoryPool::increaseNumOfAvailableChunks() {
+  numOfAvailableChunks_++;
+  metrics_.increaseAvailableChunksNum();
+}
+
+void RawMemoryPool::decreaseNumOfAllocatedChunks() {
+  numOfAllocatedChunks_--;
+  metrics_.decreaseAllocatedChunksNum();
+}
+
+void RawMemoryPool::decreaseNumOfAvailableChunks() {
+  numOfAvailableChunks_--;
+  metrics_.decreaseAvailableChunksNum();
+}
+
+char* RawMemoryPool::allocateChunk() {
+  char* chunk = nullptr;
+  if (numOfAllocatedChunks_ < maxChunksNum_) {
+    chunk = new char[chunkSize_];
+    increaseNumOfAllocatedChunks();
+    increaseNumOfAvailableChunks();
+  }
+  if (chunk) LOG_DEBUG(logger(), "A chunk has been allocated" << KVLOG(numOfAllocatedChunks_, numOfAvailableChunks_));
```

Comment:
I think it's nice also to have log in case maxChunksNum_ is reached and a chunk wasn't allocated

Reply:
OK